### PR TITLE
Mh/skate-design-patterns-and-composition

### PR DESCRIPTION
**Pull request by Hotell**
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/wc-catalogue/blaze-elements/blob/master/docs/CONTRIBUTING.md#commit
- [ ] There are no linting errors and code is properly formatted (your run before push `yarn ts:format && yarn ts:lint:fix`)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")
```
[ ] Bugfix
[x] Feature
[x] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)
- We use color and disabled props for every component - not DRY
- Styles are not scoped on non native shadow dom browsers

**What is the new behavior?**
- scoped styles cross browser via `Css` Mixin
- rendering css via `@renderCss` decorator
- `@colored` and `@disabled` decorators for alternative props extension
- mixins for `Colored` and `Disabled` behaviour
- `deco` decorators for `@bind` et all
-  TS 2.2

**Does this PR introduce a breaking change?** (check one with "x")
```
[ ] Yes
[x] No
```

**Other information**:
Team pls, check it out and ask any questions you have!

Build is failing because of https://github.com/skatejs/skatejs/pull/1085

### CODE REVIEW

File: global.d.ts [hunk 1]
```diff
@@ -1,10 +1,49 @@
...
+  const _:string;
+  export default _;
+};
 declare module '*.json';
 
 // Skate
 declare module 'core-js';
 
+// Decko
+declare module 'decko' {
```
Comment (hopkins-tk) [inline global.d.ts:15]: `decko`? :smile:

File: package.json [hunk 2]
```diff
@@ -65,7 +68,7 @@
     "tslint": "^4.4.2",
     "tslint-loader": "^3.3.0",
     "tslint-react": "^2.3.0",
-    "typescript": "^2.1.4",
+    "typescript": "rc",
```
Comment (hopkins-tk) [inline package.json]: Shouldn't there be: `"typescript": "2.2.0",` ?

File: global.d.ts (hunk 1 above)
Comment (Hotell) [reply global.d.ts:15]: image

What's your problem dude ?! :D

File: packages/_helpers/mixins.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+
+
+export function Css<BC extends Constructable<{}>>( Base: BC) {
+  return class extends Base {
+      readonly css: string;
+
+      get shadyCss() {
+        if (IS_DEV && !('css' in this)) {
+          throw new Error(`you have to implement 'css' property when using 'Css' Mixin!`);
```
Comment (hopkins-tk) [inline packages/_helpers/mixins.ts]: Shouldn't the exception be thrown always?
- especially in tests - or are those covered by that?

File: packages/button/index.demo.tsx [hunk 4]
```diff
@@ -1,22 +1,39 @@
 import { h, Component, prop, props, ComponentProps } from 'skatejs';
+import { Css, customElement, renderCss } from '../_helpers';
 import { Button } from './index';
 
 export type DemoProps = { logger: string[] };
-export class Demo extends Component<DemoProps> {
-  static get is() { return 'bl-button-demo'; }
+
+@customElement( 'bl-button-demo' )
```
Comment (hopkins-tk) [inline packages/button/index.demo.tsx]: I really like this - you probably know why :-D

File: packages/button/Button.tsx [hunk 5]
```diff
@@ -35,44 +40,40 @@ declare global {
   }
 }
 
-export class Button extends Component<ButtonProps> {
+export class Button extends Colored( Disabled( Css( Component ) ) )<ButtonProps> {
```
Comment (hopkins-tk) [inline packages/button/Button.tsx]: This is a bit cumbersome, wouldn't it be possible to use decorators for denoting that?

Wouldn't it be more readable?
- What do you think...

File: packages/button/Button.tsx [hunk 6]
```diff
@@ -1,24 +1,29 @@
...
 
 export type Props = {
-  disabled?: boolean,
   block?: boolean,
   close?: boolean,
   ghost?: boolean,
-  color?: ColorType,
-};
+}
+& ColoredProps
```
Comment (hopkins-tk) [inline packages/button/Button.tsx:24]: It is pity that we have to kind of duplicate ourselves:
- add Mixin
- by the way the props are containing those from Mixin :-)

File: packages/_helpers/decorators.ts [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+    Object.defineProperty(
+      Ctor,
+      'props',
+      {
+        configurable: true,
+        get() { return newProps; }
+      }
+    );
+
+    // this is needed for Babel ... :-/
```
Comment (hopkins-tk) [inline packages/_helpers/decorators.ts:59]: Why? - is there any issue, you could link to the comment?

File: packages/_helpers/decorators.ts [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+        value( this: { shadyCss: string }, ...args: any[] ) {
+          // tslint:disable-next-line:no-invalid-this
+          return [ h( 'style', this.shadyCss ) ].concat( originalRenderCallback.apply( this, args ) );
+        }
+      }
+    } );
+
+    return Target;
+
+    // return class extends Target {
```
Comment (hopkins-tk) [inline packages/_helpers/decorators.ts:190]: Comments should be removed before merging ;-)

File: packages/tag/TagSelector.tsx [hunk 9]
```diff
@@ -14,30 +18,18 @@ interface TagSelectorProps {
...
-      }),
-      delimiter: prop.string( {
-        attribute: true
-      }),
-    };
-  }
 
-  tags: string[] = [];
-  delimiter = ' '; // default value is space ' '
+  @prop({
```
Comment (hopkins-tk) [inline packages/tag/TagSelector.tsx:22]: Great ;-)

File: package.json (hunk 2 above)
Comment (Hotell) [reply package.json]: Bah fkn yarn... Thx

File: packages/_helpers/decorators.ts (hunk 7 above)
Comment (Hotell) [reply packages/_helpers/decorators.ts:59]: Babel is using obsolete version of decorator proposal. Even default values doesn't work... One of many reasons why I prefer TS as transpiler

File: packages/button/Button.tsx (hunk 6 above)
Comment (Hotell) [reply packages/button/Button.tsx:24]: Hmm I don't follow ?

File: packages/_helpers/decorators.ts (hunk 8 above)
Comment (Hotell) [reply packages/_helpers/decorators.ts:190]: Yup. I have to rebase and format and cleanup after team review) discussion. I don't like to add one reviewer for big feature ( like with app layout for instance ehm )

**Review by devopsmariocom — APPROVED** [review]

File: packages/_helpers/decorators.ts (hunk 8 above)
Comment (hopkins-tk) [reply packages/_helpers/decorators.ts:190]: Still not removed ;-)

**Pull request merged**